app/stable_appconfig.go

```diff
@@ -59,7 +68,102 @@ func setCustomOrderEndBlocker() []string {
 	return []string{}
 }
 
-func (app *RegenApp) registerUpgradeHandlers() {}
+func (app *RegenApp) registerUpgradeHandlers() {
+
+	// mainnet upgrade handler
+	const upgradeName = "v4.0.0"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// run state migrations for sdk modules
+		toVersion, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+		if err != nil {
+			return nil, err
+		}
+
+		// run x/ecocredit state migrations
+		if err := app.smm.RunMigrations(ctx, app.AppCodec()); err != nil {
+			return nil, err
+		}
+		toVersion[ecocredit.ModuleName] = ecocreditmodule.Module{}.ConsensusVersion()
+
+		if err := recoverFunds(ctx, app.AccountKeeper, app.BankKeeper); err != nil {
+			return nil, err
+		}
+
+		return toVersion, nil
+	})
+
+	// redwood upgrade handler
+	const redwoodUpgradeName = "v4.0.0-redwood"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// run state migrations for sdk modules
+		toVersion, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+		if err != nil {
+			return nil, err
+		}
+
+		// run x/ecocredit state migrations
+		if err := app.smm.RunMigrations(ctx, app.AppCodec()); err != nil {
+			return nil, err
+		}
+		toVersion[ecocredit.ModuleName] = ecocreditmodule.Module{}.ConsensusVersion()
+
+		return toVersion, nil
+	})
+}
+
+func recoverFunds(ctx sdk.Context, ak authkeeper.AccountKeeper, bk bankkeeper.Keeper) error {
+	// address with funds inaccessible
+	const lostAddr = "regen1c3lpjaq0ytdtsrnjqzmtj3hceavl8fe2vtkj7f"
+	// address that the community member has access to
+	const newAddr = "regen14tpuqrwf95evu3ejm9z7dn20ttcyzqy3jjpfv4"
```

Comment:
cc @clevinson @dpdanpittman

Reply:
Following up on this... @ryanchristo and I just spent some time reviewing this PR in more detail (in particular the `recoverFunds()` implementation), and I think this may work as desired even if the user has a pre-existing balance.

The error when setting a periodic vesting account on a pre-existing account only happens in the message server for [`CreateVestingAccount()`](https://github.com/cosmos/cosmos-sdk/blob/main/x/auth/vesting/msg_server.go#L52-L54).

So I think the way we are doing it (by ensuring that the vesting acc balance is added to the pre-existing new account balance), and then calling `accountKeeper.SetAccount()` should avoid an issues.

Reply:
See https://github.com/regen-network/regen-ledger/pull/1121#discussion_r888786845